File: regions/system/library/go/building-blocks/component.go
package buildingblocks

import (
	"context"
)

// ComponentStatus represents the current state of a component.
type ComponentStatus string

const (
	StatusUninitialized ComponentStatus = "uninitialized"
	StatusReady         ComponentStatus = "ready"
	StatusDegraded      ComponentStatus = "degraded"
	StatusClosed        ComponentStatus = "closed"
	StatusError         ComponentStatus = "error"
)

// Metadata holds component metadata.
type Metadata struct {
	Name    string            `json:"name"`
	Version string            `json:"version"`
	Tags    map[string]string `json:"tags,omitempty"`
}

// Component is the base interface for all building block components.
type Component interface {
	Name() string
	Version() string
	Init(ctx context.Context, metadata Metadata) error
	Close(ctx context.Context) error
	Status(ctx context.Context) ComponentStatus
}

File: regions/system/library/typescript/building-blocks/src/component.ts
export type ComponentStatus = 'uninitialized' | 'ready' | 'degraded' | 'closed' | 'error';

export interface Component {
  readonly name: string;
  readonly componentType: string;
  init(): Promise<void>;
  close(): Promise<void>;
  status(): Promise<ComponentStatus>;
  metadata(): Record<string, string>;
}
